examples/optim.jl

```diff
@@ -0,0 +1,299 @@
+          n -= 1
+          n == 0 && return
+      end
+  end
+end
+
+update!(agent, model) = agent.status == :I && (agent.days_infected += 1)
+
+function recover_or_die!(agent, model)
+  if agent.days_infected ≥ model.infection_period
+      if rand() ≤ model.death_rate
+          kill_agent!(agent, model)
+      else
+          agent.status = :R
+          agent.days_infected = 0
+      end
+  end
+end
+
+# Now we need to define a cost function. The cost function takes as agruments the model parameters that we want to tune, here migration rate, death rate, transmission rate when an infected person has been (not) detected (`β_det`, `β_und`), infection period, reinfection probability, and time until the infection is detected. The function returns one or more numbers as the objective to be minimized. Here, we try to minimize the number of infected people after 50 days.
+
+using BlackBoxOptim
+import Statistics: mean
+
+function cost(x)
+  migration_rate, death_rate, β_det, β_und, infection_period, reinfection_probability, detection_time = x
+  C = 3
+  params = create_params(C=C,
+  Ns=[500 for i in 1:C],
+  β_det=[β_det for i in 1:C],
+  migration_rate=migration_rate,
+  infection_period = infection_period,
+  reinfection_probability = reinfection_probability,
+  detection_time = detection_time,
+  death_rate = death_rate,
+  Is = ones(Int, C),
+  β_und = [β_und for i in 1:C]
+  )
+
+  model = model_initiation(; params...)
+
+  infected_fraction(model) = count(a.status == :I for a in values(model.agents)) / nagents(model)
+  _, data = run!(model, agent_step!, 50; mdata = [infected_fraction], when_model=[50], replicates=10)
+
+  return mean(data.infected_fraction)
+end
+
+# Because ABMs are stochastic, we run 10 replicates and take the average fraction of infected people after 50 days 
+
+# We can now test the function cost with some reasonable parameter values.
+
+Random.seed!(10)
+
+migration_rate=0.2
+death_rate=0.1
+β_det = 0.05
+β_und = 0.3
+infection_period = 10
+reinfection_probability = 0.1
+detection_time = 5
+x0 = [migration_rate, death_rate, β_det, β_und, infection_period, reinfection_probability, detection_time]
+cost(x0)
+
+# After 50 days, 94% of the population is infected.
+
+# We let the optimization algorithm change parameters to minimize the number of infected individuals. Note that we can limit the allowed range for each parameter separately.
+
+result = bboptimize(cost, SearchRange = [(0.0, 1.0), (0.0, 1.0), (0.0, 1.0), (0.0, 1.0), (7.0, 13.0),(0.0, 1.0),(2.0, 6.0)], NumDimensions = 7, MaxTime=20)
+best_fitness(result)
+
+# The fraction of the infected is down to 11%. Parameter values that give this result are:
+
+best_candidate(result)
+
+# We notice that the death rate is 96%, and transmission rates have also increased, while reinfection probability is much smaller. When all the infected indiduals die, infection doesn't transmit. Let's modify the cost function to give us a better solution that doesn't involve so much death.
```

Comment:
Let's modify the cost function to give us a better solution that doesn't involve so much death.
-> Let's modify the cost function to also keep the mortality rate low.